google-cloud-bigquery/src/main/scala/akka/stream/alpakka/googlecloud/bigquery/scaladsl/BigQueryQueries.scala

```diff
@@ -0,0 +1,164 @@
+
+  /**
+   * Runs a BigQuery SQL query.
+   * @see [[https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/query BigQuery reference]]
+   *
+   * @param query a query string, following the BigQuery query syntax, of the query to execute
+   * @param dryRun if set to `true` BigQuery doesn't run the job and instead returns statistics about the job such as how many bytes would be processed
+   * @param useLegacySql specifies whether to use BigQuery's legacy SQL dialect for this query
+   * @tparam Out the data model of the query results
+   * @return a [[akka.stream.scaladsl.Source]] that emits an [[Out]] for each row of the result and materializes
+   *         a [[scala.concurrent.Future]] containing the [[akka.stream.alpakka.googlecloud.bigquery.model.QueryJsonProtocol.QueryResponse]]
+   */
+  def query[Out](query: String, dryRun: Boolean = false, useLegacySql: Boolean = true)(
+      implicit um: FromEntityUnmarshaller[QueryResponse[Out]]
+  ): Source[Out, Future[QueryResponse[Out]]] = {
+    val request = QueryRequest(query, None, None, None, Some(dryRun), Some(useLegacySql), None)
+    this.query(request).mapMaterializedValue(_._2)
+  }
+
+  /**
+   * Runs a BigQuery SQL query.
+   * @see [[https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/query BigQuery reference]]
+   *
+   * @param query the [[akka.stream.alpakka.googlecloud.bigquery.model.QueryJsonProtocol.QueryRequest]]
+   * @tparam Out the data model of the query results
+   * @return a [[akka.stream.scaladsl.Source]] that emits an [[Out]] for each row of the results and materializes
+   *         a [[scala.concurrent.Future]] containing the [[akka.stream.alpakka.googlecloud.bigquery.model.JobJsonProtocol.JobReference]] and
+   *         a [[scala.concurrent.Future]] containing the [[akka.stream.alpakka.googlecloud.bigquery.model.QueryJsonProtocol.QueryResponse]]
+   */
+  def query[Out](query: QueryRequest)(
+      implicit um: FromEntityUnmarshaller[QueryResponse[Out]]
+  ): Source[Out, (Future[JobReference], Future[QueryResponse[Out]])] =
+    Source
+      .fromMaterializer { (mat, attr) =>
+        import BigQueryException._
+        import SprayJsonSupport._
+        import mat.executionContext
+        implicit val system = mat.system
+        implicit val settings = BigQueryAttributes.resolveSettings(attr, mat)
+
+        Source.lazyFutureSource { () =>
+          for {
+            entity <- Marshal(query).to[RequestEntity]
+            initialRequest = HttpRequest(POST, BigQueryEndpoints.queries(settings.projectId), entity = entity)
+            response <- BigQueryHttp().retryRequestWithOAuth(initialRequest)
+            initialQueryResponse <- Unmarshal(response.entity).to[QueryResponse[Out]]
+          } yield {
+
+            val jobReference = initialQueryResponse.jobReference
+            val jobId = jobReference.jobId.getOrElse {
+              throw BigQueryException("Query response did not contain job id.")
```

Comment:
A response may not have a job id if the request sends with "dryRun" param. Example
```
{
  "kind": "bigquery#queryResponse",
  "schema": {
    "fields": [
      {
        "name": "other",
        "type": "STRING",
        "mode": "NULLABLE"
      }
    ]
  },
  "jobReference": {
    "projectId": "analytics",
    "location": "europe-west2"
  },
  "totalBytesProcessed": "495297590",
  "jobComplete": true,
  "cacheHit": false
}
```

Reply:
Thanks for catching this! Just pushed a fix and added test coverage.